### ISSUE TITLE
Best way to mock Github PR in order to unit test Danger plugin?

### ISSUE THREAD

**Issue by wilrnh** [danger/danger#661]
Building out a Danger plugin that heavily relies/intereacts with the PR object, and I'd like to write some tests around it. However I could not figure out the best way to mock the PR object.

Right now the boiler-plate in spec_helper.rb currently does not initialize a full PR object, which makes total sense. But I wish I could find a way to supply a fake PR so that unit tests can run with it.

Thanks!

**Comment by orta** [comment]
I think you could use something like 
```
let dm = ...
let json = JSON.parse File.read("fixture/pr/json")
allow(dm.pr).to receive(:pr_json).and_return(json)
```

before your test, all of the PR related work relies on the `pr_json`attribute

**Comment by wilrnh** [comment]
Hey thanks for the pointer! But I'm still not quite there yet.

I'm now getting ```undefined method `pr' for #<Danger::Dangerfile:0x007f8281cb3800>``` where

```
# spec_helper.rb
...
# A stubbed out Dangerfile for use in tests
def testing_dangerfile
  env = Danger::EnvironmentManager.new(testing_env)
  Danger::Dangerfile.new(env, testing_ui)
end
```
and
```
# plugin_spec.rb
...
    #
    # You should test your custom attributes and methods here
    #
    describe 'with Dangerfile' do
      before do
        @dangerfile = testing_dangerfile
        @my_plugin = @dangerfile.my_plugin
        
        json = File.read(File.dirname(__FILE__) + '/support/fixtures/github_pr.json')
        allow(@my_plugin.pr).to receive(:pr_json).and_return(json)
      end

      it "should not shit itself" do
        @plugin.run
        expect("test").to eq("test")
      end

    end
  end
end
```

I did some digging and found the method `pr?` in dangerfile.rb, however, if I stub that, my tests still return github.pr_author (example) as nil. Could you throw me some more bones here?

**Comment by orta** [comment]
sorry, I didn't look at the reference, the `pr_json` is on the `github` object  - so try that 👍

**Comment by wilrnh** [comment]
It works! For reference for anyone else:

```
# plugin_spec.rb
...
    #
    # You should test your custom attributes and methods here
    #
    describe 'with Dangerfile' do
      before do
        @dangerfile = testing_dangerfile
        @my_plugin = @dangerfile.my_plugin
        
        json = File.read(File.dirname(__FILE__) + '/support/fixtures/github_pr.json')
        allow(@my_plugin.github).to receive(:pr_json).and_return(json)
      end

      it "should not shit itself" do
        @my_plugin.run
        expect("test").to eq("test")
      end

    end
  end
end
```

**Comment by orta** [comment]
hah - awesome @wilrnh - want to see if you can get this in the plugin template somehow? https://github.com/danger/danger-plugin-template/

**Comment by wilrnh** [comment]
Good call, will PR tonight

**Comment by wilrnh** [comment]
Uhoh, it looks like a couple methods get lost in space when mocking `github` this way:

```
# plugin.rb
puts github.pr_title
```
`rspec` spits out:
```
NoMethodError: undefined method `[]' for nil:NilClass
```

`pr_author` works fine. The only difference I can see is that `pr_title` is using `@github.pr_json["title"].to_s` while `pr_author` is using `pr_json["user"]["login"].to_s`

This isn't too major because I can still do `puts github.pr_json["title"]` and it works via both `rspec` and `danger`. But it would be nice if the mocking worked throughout. Got any other bones?

**Comment by orta** [comment]
Yep: Consult the [source code](https://github.com/danger/danger/blob/master/lib/danger/danger_core/plugins/dangerfile_github_plugin.rb) for the GitHub plugin

**Comment by wilrnh** [comment]
ps. looks like the attributes issue was fixed in v2.0.0
